cura/Arrange.py

```diff
@@ -0,0 +1,172 @@
+from UM.Scene.Iterator.DepthFirstIterator import DepthFirstIterator
+from UM.Logger import Logger
+from cura.ShapeArray import ShapeArray
+
+from collections import namedtuple
+
+import numpy
+import copy
+
+
+##  Return object for  bestSpot
+LocationSuggestion = namedtuple("LocationSuggestion", ["x", "y", "penalty_points", "priority"])
+
+##  The Arrange classed is used together with ShapeArray. Use it to find
+#   good locations for objects that you try to put on a build place.
+#   Different priority schemes can be defined so it alters the behavior while using
+#   the same logic.
+class Arrange:
+    def __init__(self, x, y, offset_x, offset_y, scale=1):
+        self.shape = (y, x)
+        self._priority = numpy.zeros((x, y), dtype=numpy.int32)
+        self._priority_unique_values = []
+        self._occupied = numpy.zeros((x, y), dtype=numpy.int32)
+        self._scale = scale  # convert input coordinates to arrange coordinates
+        self._offset_x = offset_x
+        self._offset_y = offset_y
+
+    ##  Helper to create an Arranger instance
+    #
+    #   Either fill in scene_root and create will find all sliceable nodes by itself,
+    #   or use fixed_nodes to provide the nodes yourself.
+    #   \param scene_root   Root for finding all scene nodes
+    #   \param fixed_nodes  Scene nodes to be placed
+    @classmethod
+    def create(cls, scene_root = None, fixed_nodes = None, scale = 0.5):
+        arranger = Arrange(220, 220, 110, 110, scale = scale)
```

Comment:
I thought about that, but the exact dimensions do not matter here, because we try to center everything. Perhaps in the future if everybody wants to place objects near the rear wall...